### PR TITLE
Dispose streams that don't fit in pool

### PR DESCRIPTION
**Pull request by JamesNK**
A gracefully completed HttpStream will not be disposed if there isn't room in the pool.

The only thing stream dispose does is ensure the `Http2OutputProducer.ProcessDataWrites` loop method exits. This isn't important because it will exit anyway when garbage collection happens.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 1]
```diff
@@ -680,7 +680,10 @@ private void ReturnStream(Http2Stream stream)
                 stream.DrainExpirationTicks = SystemClock.UtcNowTicks + StreamPoolExpiryTicks;
 
                 StreamPool.Push(stream);
+                return true;
             }
+
+            return false;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:686]: Why not just dispose it here?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:686]: Having a flag on ReturnStream that indicates whether it was successful or not felt better to me 🤷

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:686]: Feels over engineered 😏. This pool is literally for streams only

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:686]: I also slightly prefer disposing here. If we do, we should just blindly call `ReturnStream` and check `CanReuse` before disposing here too.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:686]: Yea, lets move both checks inside here.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: I'm going to remove ReturnStream altogether. It is only used in one place, and RemoveStream would just turn into:
```cs
_streams.Remove(stream);
ReturnStream(stream);
```
Which seems kind of pointless.

**Pull request merged**